File: src/helpers/day02.rs
#[derive(Debug)]
pub struct Bag {
    blue: i32,
    red: i32,
    green: i32,
}

impl Bag {
    pub fn new(blue: i32, red: i32, green: i32) -> Bag {
        Bag { blue, red, green }
    }

    pub fn from_string(s: &str) -> Result<Bag, &'static str> {
        let mut blue = 0;
        let mut red = 0;
        let mut green = 0;

        for part in s.split(',') {
            let parts: Vec<&str> = part.trim().split_whitespace().collect();
            if parts.len() != 2 {
                return Err("Invalid input format");
            }

            let count = parts[0].parse::<i32>().map_err(|_| "Invalid count")?;
            match parts[1] {
                "blue" => blue = count,
                "red" => red = count,
                "green" => green = count,
                _ => return Err("Unknown color"),
            }
        }

        Ok(Bag::new(blue, red, green))
    }

    // used in part 1 of the problem
    pub fn can_contain(&self, other: &Bag) -> bool {
        self.blue >= other.blue && self.red >= other.red && self.green >= other.green
    }

    // used in part 2 of the problem
    pub fn power(&self) -> i32 {
        self.blue * self.green * self.red
    }
}

#[derive(Debug)]
pub struct Game {
    pub id: i32,
    bags: Vec<Bag>,
}

impl Game {
    pub fn from_string(s: &str) -> Result<Game, &'static str> {
        let parts: Vec<&str> = s.split(':').collect();
        if parts.len() != 2 {
            return Err("Invalid input format");
        }

        let id = parts[0]
            .split_whitespace()
            .last()
            .ok_or("Invalid ID")?
            .parse::<i32>()
            .map_err(|_| "ID is not a number")?;

        let bag_parts = parts[1].split(';');
        let mut bags = Vec::new();

        for part in bag_parts {
            match Bag::from_string(part) {
                Ok(bag) => bags.push(bag),
                Err(e) => return Err(e),
            }
        }

        Ok(Game { id, bags })
    }

    // used in part 1 of the problem
    pub fn check(&self, max_bag: &Bag) -> bool {
        self.bags.iter().all(|bag| max_bag.can_contain(bag))
    }

    // used in part 2 of the problem
    pub fn max_values(&self) -> Bag {
        let mut max_blue = 0;
        let mut max_red = 0;
        let mut max_green = 0;

        for bag in &self.bags {
            max_blue = max_blue.max(bag.blue);
            max_red = max_red.max(bag.red);
            max_green = max_green.max(bag.green);
        }

        Bag::new(max_blue, max_red, max_green)
    }
}

File: python/aoc_2023/day02.py

from dataclasses import dataclass
from typing import List, Optional


@dataclass
class Bag:
    blue: Optional[int] = 0
    red: Optional[int] = 0
    green: Optional[int] = 0

    @classmethod
    def from_string(cls, s: str):
        "input should look something like: 3 blue, 4 red, 2 green"
        d = {}
        for part in s.strip().split(','):
            count, color = part.split()
            d[color] = int(count)
        return cls(**d)

    def can_contain(self, other: "Bag"):
        return self.red >= other.red and self.green >= other.green and self.blue >= other.blue
    
    def power(self) -> int: 
        return self.red * self.green * self.blue
    

@dataclass
class Game:
    id: int
    bags: List[Bag]

    @classmethod
    def from_string(cls, s: str):
        "input sould look something like: Game 1: 3 blue; ..."
        id_part, bag_parts = s.split(':', 1)
        id = id_part.split()[-1]
        bags = []
        for part in bag_parts.split(';'):
            bag = Bag.from_string(part)
            bags.append(bag)
        return cls(id=int(id), bags=bags)

    def check(self, max_bag: Bag):
        for bag in self.bags:
            if not max_bag.can_contain(bag):
                return False
        return True    
    
    def max_values(self) -> Bag:
        max_red = max_green = max_blue = 0
        for bag in self.bags:
            max_red = max(max_red, bag.red)
            max_green = max(max_green, bag.green)
            max_blue = max(max_blue, bag.blue)
        return Bag(red=max_red, green=max_green, blue=max_blue)